okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+                                }
+                            }
+                            return null;
+                        }
+
+                        @Override
+                        public Void visit(QosException.Unavailable exception) {
+                            Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                            if (!backoff.isPresent()) {
+                                log.debug("Max number of retries exceeded, failing call");
+                                callback.onFailure(call, new IOException("Failed to complete the request due to a "
+                                        + "server-side QoS condition: 503", exception));
+                            } else {
+                                log.debug("Rescheduling call after backoff",
+                                        SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                                scheduleExecution(callback, backoff.get());
+                            }
+                            return null;
+                        }
+                    });
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
+                if (ioException.isPresent()) {
+                    callback.onFailure(call, ioException.get());
+                    return;
+                }
+
+                callback.onFailure(call, new IOException("Failed to handle request, this is an http-remoting bug."));
+            }
+        });
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
+        enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                future.setException(exception);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.set(response);
+            }
+        });
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Call was interrupted during execution");
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof IoRemoteException) {
+                throw ((IoRemoteException) e.getCause()).getWrappedException();
```

Comment:
I think this is a bit sneaky - all 10 other exceptions thrown in this file are IOExceptions but this one is a RuntimeException.

I think I'd prefer this unwrapping to happen conspicuously out in the RemotingOkHttpClient, so that it can be clearly called out as the sync equivalent of the stuff in AsyncSerializableErrorCallAdapterFactory.

As a minimum I think your IoRemoteException really needs some explanatory [javadoc](https://github.com/palantir/http-remoting/pull/623/files#diff-413ccc6ce7a71f7570f59237fcf103b7R25) to describe why we're wrapping things

Reply:
yep, fair. will tackle this in later PR to keep this one somewhat contained. https://github.com/palantir/http-remoting/issues/626